Extend guess range outside of data range in Muon analysis

Harrietbrown opened the pull request:
**Description of work.**

This PR adds the ability to plot the fit guess outside of the range of the data in muon analysis.

**To test:**
1. Open muon analysis.
2. load MUSR 62260-3
3. go to fitting
4. add gausosc function
5. check plot guess by default it plots using the same range as the data
6. in the fit options change the `plot guess using` to x at data points. this allows you to chane the number of points to be used in the guess workspace
7. chane the number of points and the plot will update.
8. change the `plot guess using` to custom x range, this allows you to change the start_x and end_x as well as the number of points in the guess
9. change the start_x and end_x and see the plot update.
10. fit the function
11. got to results and create a results table,
12. go to model fitting, and repete the same steps as for the fitting tab. all the plot guess options should work the same.


<!-- Instructions for testing. -->

Fixes #31876 


---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Harrietbrown committed:
Add options for custom fit range for guesses in muon

This commit adds methonds for having fit guesses outside of the normal range. still to do is having default initation of the model, currently if you don't manually set things you get an error.

re: #31876

Harrietbrown committed:
Make fit guess obtain settings from fit options browser

This commit makes the fit guess in muons fit using parameters in the fit options browser. This fixes an issue in the previous commit that caused a crash if the guess parameters had not been updated when the plot guess was clicked.
This pr also hides guess parameter rows when not required.

re: #31876

Harrietbrown committed:
fix failing unit tests and remove tmp workspace.

re: #31876

Harrietbrown committed:
Add unit tests for the plot guess changes

re: 31876

AnthonyLim23 commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/fitting_widgets/basic_fitting/basic_fitting_model.py:
```diff
@@ -783,18 +814,36 @@ def _evaluate_plot_guess(self, plot_guess: bool) -> str:
         except RuntimeError:
             logger.error("Failed to plot guess.")
             return ""
+        AnalysisDataService.remove('__tmp_guess_workspace')
```
https://github.com/mantidproject/mantid/blob/main/qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/ADSHandler/ADS_calls.py#L15

AnthonyLim23 replied:
I also think the users want to be able to "save" the result of their guess to the ADS

Harrietbrown replied:
I've realised that this causes a loop by triggering a signal when ADS is modified. for the moment I have stopped the removal of the tmp workspace.

Harrietbrown replied:
The __tmp_guess_workspace does not contain the actual guess, that is in the output of the evaluate function. this tmp workspace is just a dummy with the right x range to fit over

AnthonyLim23 replied:
ah ok, could we not just keep it internally and not add it to the ADS?

AnthonyLim23 commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/fitting_widgets/basic_fitting/basic_fitting_model.py:
```diff
@@ -783,18 +814,36 @@ def _evaluate_plot_guess(self, plot_guess: bool) -> str:
+        if self.fitting_context.plot_guess_type == 'x from plot range':
+            start_x = self.current_start_x
+            end_x = self.current_end_x
+            points = 250
```
why 250

Harrietbrown replied:
This is a holdover for now, I am in the middle of checking what the most effective way of using the parameters from the workspace

AnthonyLim23 replied:
fair enough. I would assume this would mean the same number of data points as my data, just evenly spaced

Harrietbrown replied:
I was assuming it was supposed to be the same data range and number of points as the origional data

AnthonyLim23 replied:
yep, just evenly spaced across that range

Harrietbrown committed:
Have evaluateFunctino use the data workspace when x form data range.

re: #31876

Harrietbrown committed:
Remove tmp workspace from guess fit

This commit removes the tmp workaspace form the ads for the process of plotting the fit guess.

re: #31876

Harrietbrown committed:
change plot guess types to be in line with user requests.

re: #31876

AnthonyLim23 commented on docs/source/release/v6.3.0/muon.rst:
```diff
@@ -50,6 +50,7 @@ Improvements
 - The plots no longer use scientific notation for the axis values.
 - On resizing priority is given to plotting.
 - The Sequentially Fit all button is now visible for 4K displays
+- the plot guess option in Fitting can now have its range extended beyond the range of the data.
```
```suggestion
- the plot guess option in Fitting can now have its range interpolated or extrapolated. 
```

AnthonyLim23 commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/fitting_widgets/basic_fitting/basic_fitting_model.py:
```diff
@@ -783,14 +814,25 @@ def _evaluate_plot_guess(self, plot_guess: bool) -> str:
+        else:
+            raise ValueError('Plot guess type: ' + self.fitting_context.plot_guess_type + ' is not recognised')
+
+        tmp_guess_workspace = CreateWorkspace(DataX=data, DataY=data, StoreInADS=False)
```
I think we should put this into the alg helpers and make `StoreInADS` an input parameter with a default of True (to be consistent with the other algs)

Harrietbrown committed:
Move create workspace to alg creator

re: #31876

robertapplin commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/fitting_widgets/basic_fitting/basic_fitting_model.py:
```diff
@@ -812,6 +856,17 @@ def _evaluate_double_pulse_function(self, fit_function: IFunction, output_worksp
         alg.setProperty("Output", output_workspace)
         alg.execute()
 
+    def _create_guess_range_workspace(self, x_range: np.array, output_workspace: str, store_in_ads: bool = True) -> Workspace:
+        """Create a workspace with the requested fit range."""
+        alg = AlgorithmManager.create("CreateWorkspace")
+        alg.initialize()
+        alg.setAlwaysStoreInADS(store_in_ads)
+        alg.setProperty("DataX", x_range)
+        alg.setProperty("DataY", x_range)
+        alg.setProperty("OutputWorkspace", output_workspace)
+        alg.execute()
+        return alg.getProperty('OutputWorkspace').value
```
Can we use this utility function which is in the algorithm utils file?
https://github.com/mantidproject/mantid/blob/cffb8050f43a522a2ff059a7834b0e4f59aa1075/qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/utilities/algorithm_utils.py#L377

Harrietbrown replied:
We don;t want te workspace stored in the ADS so I need to set that in the algorithm

AnthonyLim23 replied:
@Harrietbrown does it still work with workspace history? 

We could modify the  function to be
```
def run_create_workspace(x_data, y_data, name, store_in_ADS=True)
```

Harrietbrown replied:
I've tried it, but the function returns a workspace name not a workspace itself. so it can;t be used in the Evaluate function since it is not in the ADS.

robertapplin replied:
Would something like this work?
```
def run_create_workspace(x_data, y_data, name, store_in_ADS=True, get_object: bool = False)
```
and then
```
if get_object:
    return alg.getProperty('OutputWorkspace').value 
else:
    return alg.getProperty('OutputWorkspace').valueAsStr
```

robertapplin replied:
Or maybe we don't need `get_object` and we can just do:
```
if not store_in_ADS:
    return alg.getProperty('OutputWorkspace').value 
else:
    return alg.getProperty('OutputWorkspace').valueAsStr
```

Harrietbrown replied:
As typehints are becoming more and more used in our codebase I am hesitent to implement a change that would make a python function return different types.

robertapplin commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/fitting_widgets/basic_fitting/basic_fitting_model.py:
```diff
@@ -783,14 +814,27 @@ def _evaluate_plot_guess(self, plot_guess: bool) -> str:
+            data = np.linspace(self.fitting_context.plot_guess_start_x, self.fitting_context.plot_guess_end_x,
+                               self.fitting_context.plot_guess_points)
+        else:
+            raise ValueError('Plot guess type: ' + self.fitting_context.plot_guess_type + ' is not recognised')
```
Suggestion to use an f-string
```suggestion
            raise ValueError(f"Plot guess type '{self.fitting_context.plot_guess_type}' is not recognised.")
```

robertapplin commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/fitting_widgets/basic_fitting/basic_fitting_model.py:
```diff
@@ -783,14 +814,27 @@ def _evaluate_plot_guess(self, plot_guess: bool) -> str:
 
     def _evaluate_function(self, fit_function: IFunction, output_workspace: str) -> str:
         """Evaluate the plot guess fit function and returns the name of the resulting guess workspace."""
+        if self.fitting_context.plot_guess_type == 'x from fit range':
```
I think it would be good to create three constant variables for 'x from fit range', 'Uniform points across data range' and 'Custom x range' and but them in the `basic_fitting_context` here:
https://github.com/mantidproject/mantid/blob/cffb8050f43a522a2ff059a7834b0e4f59aa1075/qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/contexts/fitting_contexts/basic_fitting_context.py#L10

Then import them to this file and do something like this for each if clause
```
if self.fitting_context.plot_guess_type == PLOT_GUESS_FIT_RANGE:
```

robertapplin commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/contexts/fitting_contexts/basic_fitting_context.py:
```diff
@@ -37,6 +37,10 @@ def __init__(self, allow_double_pulse_fitting: bool = False):
         self._chi_squared_for_undo: list = []
 
         self._plot_guess: bool = False
+        self._plot_guess_type: str = None
```
Should this have a default value? For instance we might want it defaulted to `PLOT_GUESS_FIT_RANGE` (see comment below)

robertapplin commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/fitting_widgets/basic_fitting/basic_fitting_model.py:
```diff
@@ -783,14 +814,27 @@ def _evaluate_plot_guess(self, plot_guess: bool) -> str:
+        else:
+            raise ValueError('Plot guess type: ' + self.fitting_context.plot_guess_type + ' is not recognised')
+
+        tmp_guess_workspace = self._create_guess_range_workspace(x_range=data, output_workspace='guess_range',
```
Can this variable be renamed to something like `extrapolated_workspace`/`extended_workspace`/`augmented_workspace` or something similar?

robertapplin commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/fitting_widgets/basic_fitting/basic_fitting_presenter.py:
```diff
@@ -233,6 +238,39 @@ def handle_evaluation_type_changed(self) -> None:
+    def handle_plot_guess_type_changed(self) -> None:
+        """Handle when the evaluation type is changed."""
+        self.model.plot_guess_type = self.view.plot_guess_type
+        if self.model.plot_guess_type == 'x from fit range':
```
Can these strings ('x from fit range', etc.) be imported from the basic fitting context?

robertapplin commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/fitting_widgets/basic_fitting/basic_fitting_presenter.py:
```diff
@@ -411,6 +451,12 @@ def update_plot_guess(self) -> None:
         self.model.update_plot_guess()
         self.update_plot_guess_notifier.notify_subscribers()
 
+    def update_guess_parameters(self) -> None:
```
This function is missing a comment similar to the other comments in this file

robertapplin commented on docs/source/release/v6.3.0/muon.rst:
```diff
@@ -50,6 +50,7 @@ Improvements
 - The plots no longer use scientific notation for the axis values.
 - On resizing priority is given to plotting.
 - The Sequentially Fit all button is now visible for 4K displays
+- the plot guess option in Fitting can now have its range interpolated or extrapolated.
```
'The' is missing a capital letter

robertapplin reviewed (changes requested):
The code added in this PR is very clean and easy to read, so nice work. I had a few other minor suggestions that could help this further

robertapplin commented:
I've noticed that the Plot guess custom range gets reset in the Fit Options when you change to a different dataset using `>>`, but the range of the guess on the graph stays the same

Harrietbrown committed:
Apply changes from code review

re: #31876

Harrietbrown committed:
Move workspace creator to algorithm_utils

Harrietbrown committed:
fix failing unittest

robertapplin reviewed (approved):
The code looks very good and this works well :shipit:

DavidFair reviewed (approved):
LGTM, nice use of MVP. Was a breeze to review both the tests and code :+1:
